apps/omg_watcher/lib/exit_processor/core.ex

```diff
@@ -456,27 +613,53 @@ defmodule OMG.Watcher.ExitProcessor.Core do
       competing_txbytes: raw_known_tx |> Transaction.encode(),
       competing_input_index: competing_input_index,
       competing_sig: competing_sig,
-      competing_txid: nil,
-      competing_proof: nil
+      competing_txid: known_tx_utxo_pos,
+      competing_proof: maybe_calculate_proof(known_tx_utxo_pos, blocks)
     }
   end
 
-  defp with_competitors_from_tx_appendix(%__MODULE__{in_flight_exits: ifes} = state) do
-    known_txs = get_known_txs(state)
+  defp prepare_canonical_response(ife_txbytes, known_tx_utxo_pos, blocks) do
+    %{
+      inflight_txbytes: ife_txbytes,
+      inflight_txid: known_tx_utxo_pos,
+      inflight_proof: maybe_calculate_proof(known_tx_utxo_pos, blocks)
+    }
+  end
 
-    ifes
-    |> Map.values()
-    # TODO: not enough - must take oldest competitor into account (?)
-    |> Stream.filter(&InFlightExitInfo.is_canonical?/1)
-    |> Stream.map(fn %InFlightExitInfo{tx: tx} -> tx end)
-    # TODO: expensive!
-    |> Stream.filter(fn tx -> known_txs |> Enum.find(&competitor_for(tx, &1)) end)
-    |> Stream.map(fn %{raw_tx: raw_tx} -> Transaction.encode(raw_tx) end)
+  defp maybe_calculate_proof(nil, _), do: nil
+
+  defp maybe_calculate_proof(Utxo.position(blknum, txindex, _), blocks) do
+    blocks
+    |> Enum.find(fn %Block{number: number} -> blknum == number end)
+    |> Map.fetch!(:transactions)
+    |> Enum.map(fn encoded_signed_tx ->
+      %Transaction.Recovered{tx_hash: hash} = recover_correct_tx_from_block(encoded_signed_tx)
+      hash
+    end)
+    |> Block.create_tx_proof(txindex)
+  end
+
+  defp find_competitor(known_txs, signed_ife_tx) do
+    known_txs
+    |> Enum.find(fn known -> competitor_for(signed_ife_tx, known) end)
+    |> case do
+      nil -> {:error, :competitor_not_found}
+      got_it! -> {:ok, got_it!}
+    end
   end
 
-  # tells whether a single transaction is a competitor for another single transactions, by returning nil or the
+  defp find_canonical(known_txs, raw_ife_tx) do
+    known_txs
+    |> Enum.find(fn %KnownTx{signed_tx: %Transaction.Signed{raw_tx: block_raw_tx}} -> block_raw_tx == raw_ife_tx end)
+    |> case do
+      nil -> {:error, :canonical_not_found}
+      got_it! -> {:ok, got_it!}
```

Comment:
This exclamation mark is surprising. Why not use something plain and boring like `value` instead?

Reply:
T_T, oh okaaay